src/explorer/Explorers/TextColumnExplorer.cs

```diff
@@ -52,170 +50,299 @@ select new
-                allCountsList.Add(countsList);
             }
+            return substrings;
+        }
 
-            var rand = new Random(Environment.TickCount);
-            return Enumerable.Range(0, GeneratedValuesCount * 3).Select(_
-                => GenerateString(allSubstringsList, allCountsList, rand));
+        private static async Task<bool> CheckIsEmail(DConnection conn, ExplorerContext ctx)
+        {
+            var emailCheck = await conn.Exec(
+                new TextColumnTrim(ctx.Table, ctx.Column, TextColumnTrimType.Both, EmailAddressChars));
+
+            var counts = ValueCounts.Compute(emailCheck.Rows);
+
+            return counts.TotalCount == emailCheck.Rows
+                .Where(r => r.IsNull || r.Value == "@")
+                .Sum(r => r.Count);
         }
 
-        private string GenerateString(List<StringList> substrings, List<LongList> counts, Random rand)
+        private static async Task<SubstringWithCountList> ExploreEmailDomains(DConnection conn, ExplorerContext ctx)
         {
-            var sb = new StringBuilder();
-            var len = rand.Next(3, substrings.Count);
-            for (var pos = 0; pos < substrings.Count && sb.Length < len; pos++)
+            var domains = await conn.Exec(new TextColumnTrim(ctx.Table, ctx.Column, TextColumnTrimType.Leading, EmailAddressChars));
+            var totalCount = 0L;
+            var domain = new SubstringWithCountList();
+            foreach (var row in domains.Rows)
             {
-                var str = FindSubstring(substrings[pos], counts[pos], rand);
-                sb.Append(str);
-                pos += str.Length;
+                if (row.HasValue && row.Value.StartsWith("@", StringComparison.InvariantCulture))
+                {
+                    totalCount += row.Count;
+                    domain.Add((row.Value, totalCount));
+                }
             }
-            return sb.ToString();
+            return domain;
         }
 
-        private string FindSubstring(StringList substrings, LongList counts, Random rand)
+        private static async Task<SubstringWithCountList> ExploreEmailTopLevelDomains(DConnection conn, ExplorerContext ctx)
         {
-            if (substrings.Count == 0)
+            var suffixes = await conn.Exec(new TextColumnSuffix(ctx.Table, ctx.Column, 3, 7));
+            var totalCount = 0L;
+            var tlds = new SubstringWithCountList();
+            foreach (var row in suffixes.Rows)
+            {
+                if (row.HasValue && row.Value.StartsWith(".", StringComparison.InvariantCulture))
+                {
+                    totalCount += row.Count;
+                    tlds.Add((row.Value, totalCount));
+                }
+            }
+            return tlds;
+        }
+    }
+
+    internal class SubstringWithCountList : List<(string Value, long Count)>
+    {
+        public long TotalCount => Count == 0 ? 0 : this[^1].Count;
+
```

Comment:
I would suggest adding a simple method to abstract adding a new entry (and incrementing the count):

```suggestion
        public void AddValueCount(string value, long count)
        {
            Add((value, TotalCount + count));
        }
```

This way you don't have to remember to increment the count from the calling code. 

You could even add a static initializer for creating the list from a list of `ValueWithCount` (see previous comment).

```suggestion
        public static SubstringWithCountList BuildSubstringCounts(IEnumerable<ValueWithCount<string>> valueCounts)
        {
            var sscl = new SubstringWithCountList();
            foreach (var vc in valueCounts)
            {
                sscl.AddValueCount(vc.Value, vc.Count);
            }
            return sscl;
        }
        
        public void AddValueCount(string value, long count)
        {
            Add((value, TotalCount + count));
        }
```